File: python/minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t): return ''

        window, tCnt = defaultdict(int), defaultdict(int)
        for i in range(len(t)):
            tCnt[t[i]] += 1
        l = 0
        have, need = 0, len(t)
        res, resLen = [-1, -1], float('inf')

        for r in range(len(s)):
            window[s[r]] += 1
            have += 1 if window[s[r]] <= tCnt[s[r]] else 0
            while have == need:
                if r - l + 1 < resLen:
                    res = [l, r]
                    resLen = r - l + 1
                window[s[l]] -= 1
                if s[l] in tCnt and window[s[l]] < tCnt[s[l]]:
                    have -= 1
                l += 1
        l, r = res
        return s[l:r+1] if resLen != float('inf') else ''

File: C++/minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.size() < t.size()){return "";}
        unordered_map<char, int> window; unordered_map<char, int> cntT;

        for (char c : t){
            if (!cntT.contains(c)){cntT[c] = 0;}
            cntT[c]++;
        }

        int l = 0;
        int have = 0; int need = t.size();
        pair<int, int> res = {-1, -1}; int resLen = INT_MAX;

        for (int r = 0; r < s.size(); r++){
            if(!window.contains(s[r])){window[s[r]] = 0;}
            window[s[r]]++;
            have += (window[s[r]] <= cntT[s[r]]) ? 1 : 0;

            while (have == need){
                if (r - l + 1 < resLen){
                    res.first = l; res.second = r;
                    resLen = r - l + 1;
                }
                window[s[l]]--;
                if (cntT.contains(s[l]) && window[s[l]] < cntT[s[l]]){have--;}
                l++;
            }
        }

        l = res.first; int r = res.second;
        return (resLen != INT_MAX) ? s.substr(l, r - l + 1) : "";
    }
};
